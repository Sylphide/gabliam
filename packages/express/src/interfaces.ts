import { express } from './express';

/**
 * Config for express plugin
 */
export interface ExpressPluginConfig {
  /**
   * Root path of express plugin
   */
  rootPath: string;

  /**
   * Port of express
   */
  port: number;

  /**
   * Hostname of express
   */
  hostname: string;
}

/**
 * Represent a method that create an express router
 */
export type RouterCreator = () => express.Router;
