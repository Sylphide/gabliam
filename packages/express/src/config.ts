import { Bean, Joi, PluginConfig, Value } from '@gabliam/core';
import { WebPluginConfig, WEB_PLUGIN_CONFIG } from '@gabliam/web-core';
import * as bodyParser from 'body-parser';

const options = Joi.object().keys({
  inflate: Joi.boolean(),
  limit: [Joi.number(), Joi.string()],
  type: [Joi.string(), Joi.array().items(Joi.string())],
});

const optionsJson = options.keys({
  strict: Joi.boolean(),
});

const optionsText = options.keys({
  defaultCharset: Joi.string(),
});

const optionsUrlencoded = options
  .keys({
    extended: Joi.boolean(),
    parameterLimit: Joi.number(),
  })
  .default({ extended: false });

const configJoi = Joi.object()
  .keys({
    json: Joi.boolean().default(true),
    optionsJson,
    urlencoded: Joi.boolean().default(true),
    optionsUrlencoded,
    text: Joi.boolean().default(false),
    optionsText,
  })
  .default({
    json: true,
    urlencoded: true,
    optionsUrlencoded: { extended: false },
    text: false,
  });

export interface BodyParserConfig {
  json: boolean;
  optionsJson?: bodyParser.OptionsJson;
  urlencoded: boolean;
  optionsUrlencoded?: bodyParser.OptionsUrlencoded;
  text: boolean;
  optionsText?: bodyParser.OptionsText;
}

export interface ExpressConfig extends WebPluginConfig {
  bodyParser: BodyParserConfig;
}

@PluginConfig()
export class ExpressPluginConfig implements WebPluginConfig {
  @Value('application.web.rootPath', Joi.string())
  rootPath = '/';

  @Value('application.web.port', Joi.number().positive())
  port: number = process.env.PORT ? parseInt(process.env.PORT!, 10) : 3000;

  @Value('application.web.hostname', Joi.string())
  hostname: string;

  @Value('application.web.bodyParser', configJoi)
  bodyParser: BodyParserConfig;

  @Bean(WEB_PLUGIN_CONFIG)
  restConfig(): ExpressConfig {
    return {
      rootPath: this.rootPath,
      port: this.port,
      hostname: this.hostname,
      bodyParser: this.bodyParser,
    };
  }
}
