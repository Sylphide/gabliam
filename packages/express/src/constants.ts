export const DEFAULT_ROUTING_ROOT_PATH = '/';

export const EXPRESS_PLUGIN_CONFIG = Symbol('GABLIAM/EXPRESS_PLUGIN_CONFIG');

export const APP = Symbol('GABLIAM/EXPRESS_APP');

/**
 * Represent a custom router creator
 *
 * If you want a custom router use this const for create your router
 *
 * ## Sample
 * @Config()
 * class Conf {
 *  @Bean(CUSTOM_ROUTER_CREATOR)
 *  custom() {
 *    return () =>
 *      e.Router({
 *        caseSensitive: true
 *      });
 *  }
 * }
 */
export const CUSTOM_ROUTER_CREATOR = Symbol('GABLIAM/CUSTOM_ROUTER_CREATOR');

export const SERVER = Symbol('GABLIAM/EXPRESS_SERVER');

export const ERRORS_MSGS = {
  DUPLICATED_CONFIG_DECORATOR: `Cannot apply @ExpressConfig or @ExpressErrorConfig decorator multiple times on same method.`
};
