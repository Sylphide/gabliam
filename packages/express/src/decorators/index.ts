export * from './controller';
export * from './config';
export * from './method';
