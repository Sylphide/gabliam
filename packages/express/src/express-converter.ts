import {
  Container,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  Service,
  toPromise,
} from '@gabliam/core';
import {
  createInterceptorResolver,
  extractParameters,
  getParameterMetadata,
  isInterceptor,
  InterceptorMethod,
  Interceptor,
  AfterResponseInterceptor,
} from '@gabliam/web-core';
import { express } from './express';
import { getContext } from './utils';
import { METADATA_KEY } from './constants';

@InjectContainer()
@Service()
export class ExpressConverter {
  public interceptorToMiddleware(clazz: any): express.RequestHandler {
    const container = <Container>(this as any)[INJECT_CONTAINER_KEY];
    const instance = createInterceptorResolver(container)(clazz);
    if (!isInterceptor(instance)) {
      throw new Error();
    }

    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const args = extractParameters(
        instance,
        'intercept',
        null,
        getContext(req),
        next,
        getParameterMetadata(instance, 'intercept')
      );

      try {
        await toPromise((instance['intercept'] as any)(...args));
        next();
      } catch (err) {
        next(err);
      }
    };
  }
}

export const isExpressInterceptor = (target: any) =>
  Reflect.getOwnMetadata(METADATA_KEY.expressInterceptor, target) === true;

const addExpressInterceptorMetadata = (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.expressInterceptor, true, target);
};

export const middlewareToInterceptor = (
  mid: express.RequestHandler | express.ErrorRequestHandler,
  type: InterceptorMethod
) => {
  if (type === 'intercept') {
    const clazz = class implements Interceptor {
      intercept() {
        return mid;
      }
    };
    addExpressInterceptorMetadata(clazz);
    return clazz;
  } else {
    const clazz = class implements AfterResponseInterceptor {
      afterResponse() {
        return mid;
      }
    };
    addExpressInterceptorMetadata(clazz);
    return clazz;
  }
};
