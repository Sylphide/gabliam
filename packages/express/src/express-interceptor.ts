import {
  AfterResponseInterceptor,
  Interceptor,
  InterceptorMethod,
} from '@gabliam/web-core';
import { METADATA_KEY } from './constants';
import { express } from './express';
import { Service } from '@gabliam/core';

export const isExpressInterceptor = (target: any) =>
  Reflect.getOwnMetadata(
    METADATA_KEY.expressInterceptor,
    target.constructor || target
  ) === true;

const ExpressInterceptor = () => (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.expressInterceptor, true, target);
};

export type InterceptorConstructor =
  | { new (): Interceptor }
  | { new (): AfterResponseInterceptor };

export const toInterceptor = (
  mid: express.RequestHandler | express.ErrorRequestHandler,
  type: InterceptorMethod = 'intercept'
): InterceptorConstructor => {
  let clazz: InterceptorConstructor;
  if (type === 'intercept') {
    clazz = class implements Interceptor {
      intercept() {
        return mid;
      }
    };
  } else {
    clazz = class implements AfterResponseInterceptor {
      afterResponse() {
        return mid;
      }
    };
  }
  Service()(clazz);
  ExpressInterceptor()(clazz);
  return clazz;
};
