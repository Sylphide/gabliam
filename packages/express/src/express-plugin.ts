import {
  Scan,
  Registry,
  Plugin,
  Container,
  GabliamPlugin,
  ValueExtractor,
  VALUE_EXTRACTOR
} from '@gabliam/core';
import {
  EXPRESS_PLUGIN_CONFIG,
  APP,
  SERVER,
  CUSTOM_ROUTER_CREATOR
} from './constants';
import {
  TYPE,
  METADATA_KEY,
  PARAMETER_TYPE,
  DEFAULT_PARAM_VALUE,
  getMiddlewares,
  ResponseEntity,
  ConfigMetadata,
  ControllerMetadata,
  ControllerMethodMetadata,
  ControllerParameterMetadata,
  ParameterMetadata,
  cleanPath
} from '@gabliam/rest-decorators';
import { ExpressPluginConfig, RouterCreator } from './interfaces';
import * as d from 'debug';
import * as http from 'http';
import { MiddlewareConfig, ExpressMiddlewareConfig } from './middlewares';
import { express } from './express';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Plugin('ExpressPlugin')
@Scan()
export class ExpressPlugin implements GabliamPlugin {
  /**
   * binding phase
   *
   * Bind all controller and bind express app
   * @param  {Container} container
   * @param  {Registry} registry
   */
  bind(container: Container, registry: Registry) {
    container.bind(APP).toConstantValue(express());
    registry.get(TYPE.Controller).forEach(({ id, target }) =>
      container
        .bind<any>(id)
        .to(target)
        .inSingletonScope()
    );

    container.bind(MiddlewareConfig).toConstantValue(new MiddlewareConfig());
  }

  build(container: Container, registry: Registry) {
    this.buildExpressConfig(container, registry);
    this.buildControllers(container, registry);
    this.buildExpressErrorConfig(container, registry);
  }

  /**
   * Management of @middleware decorator in config class
   *
   * @param  {Container} container
   * @param  {Registry} registry
   * @param  {any} confInstance
   */
  config(container: Container, registry: Registry, confInstance: any) {
    const middlewareConfig = container.get<ExpressMiddlewareConfig>(
      MiddlewareConfig
    );

    // if config class has a @middleware decorator, add in this.middlewares for add it in building phase
    if (
      Reflect.hasMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      )
    ) {
      const metadataList: ConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        middlewareConfig.addMiddleware({
          order,
          instance: confInstance[key].bind(confInstance[key])
        });
      });
    }

    // if config class has a @middleware decorator, add in this.errorMiddlewares for add it in building phase
    if (
      Reflect.hasMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      )
    ) {
      const metadataList: ConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        middlewareConfig.addErrorMiddleware({
          order,
          instance: confInstance[key].bind(confInstance[key])
        });
      });
    }
  }

  async destroy(container: Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>(resolve => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async start(container: Container, registry: Registry) {
    const restConfig = container.get<ExpressPluginConfig>(
      EXPRESS_PLUGIN_CONFIG
    );
    const app = container.get<express.Application>(APP);
    const port = restConfig.port;
    app.set('port', port);

    const server = http.createServer(<any>app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);
    container.bind(SERVER).toConstantValue(server);

    /* istanbul ignore next */
    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    /* istanbul ignore next */
    function onListening(): void {
      const addr = server.address();
      const bind =
        typeof addr === 'string' ? `pipe ${addr}` : `port ${addr.port}`;
      console.log(`Listening on ${bind}`);
    }
  }

  /**
   * Build express middleware
   *
   * @param  {Container} container
   * @param  {Registry} registry
   */
  private buildExpressConfig(container: Container, registry: Registry) {
    const middlewareConfig = container.get<ExpressMiddlewareConfig>(
      MiddlewareConfig
    );
    const app = container.get<express.Application>(APP);
    middlewareConfig.middlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  /**
   * Build express error middleware
   *
   * @param  {Container} container
   * @param  {Registry} registry
   */
  private buildExpressErrorConfig(container: Container, registry: Registry) {
    const middlewareConfig = container.get<ExpressMiddlewareConfig>(
      MiddlewareConfig
    );

    const app = container.get<express.Application>(APP);
    middlewareConfig.errorMiddlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  /**
   * Build all controllers
   *
   * @param  {Container} container
   * @param  {Registry} registry
   */
  private buildControllers(container: Container, registry: Registry) {
    const restConfig = container.get<ExpressPluginConfig>(
      EXPRESS_PLUGIN_CONFIG
    );

    const valueExtractor = container.get<ValueExtractor>(VALUE_EXTRACTOR);

    // get the router creator
    let routerCreator: RouterCreator = () => express.Router();
    try {
      routerCreator = container.get<RouterCreator>(CUSTOM_ROUTER_CREATOR);
    } catch (e) {}

    debug('registerControllers', TYPE.Controller);
    const controllerIds = registry.get(TYPE.Controller);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<object>(controllerId);

      const controllerMetadata: ControllerMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controller,
        controller.constructor
      );

      const controllerMiddlewares = getMiddlewares(
        container,
        controller.constructor
      );

      const methodMetadatas: ControllerMethodMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.controllerMethod,
        controller.constructor
      );

      const parameterMetadata: ControllerParameterMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controllerParameter,
        controller.constructor
      );
      // if the controller has controllerMetadata and methodMetadatas
      if (controllerMetadata && methodMetadatas) {
        const router = routerCreator();
        const controllerPath = valueExtractor(
          controllerMetadata.path,
          controllerMetadata.path
        );
        const routerPath = cleanPath(`${restConfig.rootPath}${controllerPath}`);

        debug(`New route : "${routerPath}"`);

        methodMetadatas.forEach((methodMetadata: ControllerMethodMetadata) => {
          let paramList: ParameterMetadata[] = [];
          if (parameterMetadata) {
            paramList = parameterMetadata[methodMetadata.key] || [];
          }
          let methodMetadataPath = cleanPath(
            valueExtractor(methodMetadata.path, methodMetadata.path)
          );
          if (methodMetadataPath[0] !== '/') {
            methodMetadataPath = '/' + methodMetadataPath;
          }
          const methodMiddlewares = getMiddlewares(
            container,
            controller.constructor,
            methodMetadata.key
          );
          debug(methodMetadataPath);
          debug({ methodMiddlewares, controllerMiddlewares });
          // create handler
          const handler: express.RequestHandler = this.handlerFactory(
            container,
            controllerId,
            methodMetadata.key,
            controllerMetadata.json,
            paramList
          );

          // register handler in router
          (router as any)[methodMetadata.method](
            methodMetadataPath,
            ...controllerMiddlewares,
            ...methodMiddlewares,
            handler
          );
        });
        const app = container.get<express.Application>(APP);
        app.use(routerPath, router);
      }
    });
  }

  private handlerFactory(
    container: Container,
    controllerId: any,
    key: string,
    json: boolean,
    parameterMetadata: ParameterMetadata[]
  ): express.RequestHandler {
    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const controller = container.get<any>(controllerId);
      // extract all args
      const args = this.extractParameters(
        controller,
        key,
        req,
        res,
        next,
        parameterMetadata
      );

      // response handler if the result is a ResponseEntity
      function responseEntityHandler(value: ResponseEntity) {
        if (value.hasHeader()) {
          Object.keys(value.headers).forEach(k =>
            res.setHeader(k, value.headers[k])
          );
        }
        res.status(value.status).json(value.body);
      }

      try {
        const result: any = await Promise.resolve(controller[key](...args));
        if (result !== undefined && !res.headersSent) {
          if (result instanceof ResponseEntity) {
            responseEntityHandler(result);
          } else if (json) {
            res.json(result);
          } else {
            res.send(result);
          }
        }
      } catch (err) {
        next(err);
      }
    };
  }

  private extractParameters(
    target: any,
    key: string,
    req: express.Request,
    res: express.Response,
    next: express.NextFunction,
    params: ParameterMetadata[]
  ): any[] {
    const args = [];
    if (!params || !params.length) {
      return [req, res, next];
    }

    // create de param getter
    const getParam = this.getFuncParam(target, key);

    for (const item of params) {
      switch (item.type) {
        case PARAMETER_TYPE.REQUEST:
          args[item.index] = getParam(req, null, item);
          break;
        case PARAMETER_TYPE.NEXT:
          args[item.index] = next;
          break;
        case PARAMETER_TYPE.PARAMS:
          args[item.index] = getParam(req, 'params', item);
          break;
        case PARAMETER_TYPE.QUERY:
          args[item.index] = getParam(req, 'query', item);
          break;
        case PARAMETER_TYPE.BODY:
          args[item.index] = getParam(req, 'body', item);
          break;
        case PARAMETER_TYPE.HEADERS:
          args[item.index] = getParam(req, 'headers', item);
          break;
        case PARAMETER_TYPE.COOKIES:
          args[item.index] = getParam(req, 'cookies', item);
          break;
        default:
          args[item.index] = res;
          break; // response
      }
    }
    args.push(req, res, next);
    return args;
  }

  private getFuncParam(target: any, key: string) {
    return (
      source: any,
      paramType: string | null,
      itemParam: ParameterMetadata
    ) => {
      const name = itemParam.parameterName;

      // get the param source
      let param = source;
      if (paramType !== null && source[paramType]) {
        param = source[paramType];
      }

      let res = param[name];

      if (res) {
        /**
         * For query, all value sare considered to string value.
         * If the query waits for a Number, we try to convert the value
         */
        if (paramType === 'query' || paramType === 'params') {
          const type: Function[] = Reflect.getMetadata(
            'design:paramtypes',
            target,
            key
          );
          if (Array.isArray(type) && type[itemParam.index]) {
            try {
              if (type[itemParam.index].name === 'Number') {
                // parseFloat for compatibility with integer and float
                res = Number.parseFloat(res);
              }
            } catch (e) {}
          }
        }
        return res;
      } else {
        return name === DEFAULT_PARAM_VALUE ? param : undefined;
      }
    };
  }
}
