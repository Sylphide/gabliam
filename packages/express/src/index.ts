import { APP, SERVER } from './constants';
import { ExpressPlugin } from './express-plugin';

export * from '@gabliam/rest-decorators';
export * from './middlewares';
export * from './express';
export * from './interfaces';
export * from './decorators';
export * from 'http-status-codes';
export { APP, SERVER, ExpressPlugin as default };
