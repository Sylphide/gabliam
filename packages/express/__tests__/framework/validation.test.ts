import { Joi } from '@gabliam/core';
import { Controller, Get, Post, Validate } from '@gabliam/web-core';
import * as supertest from 'supertest';
import { express as e } from '../../src';
import { ExpressPluginTest } from '../express-plugin-test';

let appTest: ExpressPluginTest;

beforeEach(async () => {
  appTest = new ExpressPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('express integration', () => {
  describe('validations', () => {
    it('req.headers', async () => {
      @Controller('/')
      class TestController {
        @Get('/')
        @Validate(
          {
            headers: {
              accept: Joi.string().regex(/xml/),
            },
          },
          { allowUnknown: true }
        )
        async getTest(req: e.Request, res: e.Response) {
          return 'lol';
        }
      }

      appTest.addClass(TestController);
      await appTest.build();
      const response = await supertest(appTest.app)
        .get('/')
        .set('Accept', 'application/json')
        .expect(400);
      expect(response).toMatchSnapshot();
    });

    it('req.params', async () => {
      @Controller('/')
      class TestController {
        @Get('/user/:id')
        @Validate({
          params: {
            id: Joi.string().token(),
          },
        })
        async getTest(req: e.Request, res: e.Response) {
          return 'lol';
        }
      }

      appTest.addClass(TestController);
      await appTest.build();
      const response = await supertest(appTest.app)
        .get('/user/@@')
        .expect(400);
      expect(response).toMatchSnapshot();
    });

    it('req.query', async () => {
      @Controller('/')
      class TestController {
        @Get('/')
        @Validate({
          query: Joi.object().keys({
            start: Joi.date(),
          }),
        })
        async getTest(req: e.Request, res: e.Response) {
          return 'lol';
        }
      }

      appTest.addClass(TestController);
      await appTest.build();
      const response = await supertest(appTest.app)
        .get('/?end=lol')
        .expect(400);
      expect(response).toMatchSnapshot();
    });

    it('req.body', async () => {
      @Controller('/')
      class TestController {
        @Post('/')
        @Validate({
          body: {
            first: Joi.string().required(),
            last: Joi.string(),
            role: Joi.number().integer(),
          },
        })
        async getTest(req: e.Request, res: e.Response) {
          return 'lol';
        }
      }

      appTest.addClass(TestController);

      await appTest.build();
      const response = await supertest(appTest.app)
        .post('/')
        .send({
          first: 'john',
          last: 123,
        })
        .expect(400);
      expect(response).toMatchSnapshot();
    });
  });

  describe('update req values', () => {
    it('req.headers', async () => {
      @Controller('/')
      class TestController {
        @Get('/')
        @Validate(
          {
            headers: {
              accept: Joi.string().regex(/json/),
              'secret-header': Joi.string().default('@@@@@@'),
            },
          },
          { allowUnknown: true }
        )
        async getTest(req: e.Request, res: e.Response) {
          delete req.headers.host; // this can change computer to computer, so just remove it
          expect(req.headers).toMatchSnapshot();
          return 'k';
        }
      }

      appTest.addClass(TestController);
      await appTest.build();
      const response = await supertest(appTest.app)
        .get('/')
        .set('Accept', 'application/json')
        .expect(200);
      expect(response).toMatchSnapshot();
    });

    it('req.params', async () => {
      @Controller('/')
      class TestController {
        @Get('/user/:id')
        @Validate({
          params: {
            id: Joi.string().uppercase(),
          },
        })
        async getTest(req: e.Request, res: e.Response) {
          expect(req.params.id).toBe('ADAM');
          return 'k';
        }
      }

      appTest.addClass(TestController);
      await appTest.build();
      const response = await supertest(appTest.app)
        .get('/user/adam')
        .expect(200);
      expect(response).toMatchSnapshot();
    });

    it('req.query', async () => {
      @Controller('/')
      class TestController {
        @Get('/')
        @Validate({
          query: Joi.object().keys({
            name: Joi.string().uppercase(),
            page: Joi.number().default(1),
          }),
        })
        async getTest(req: e.Request, res: e.Response) {
          expect(req.query).toEqual({
            name: 'JOHN',
            page: 1,
          });
          return 'k';
        }
      }

      appTest.addClass(TestController);
      await appTest.build();
      const response = await supertest(appTest.app)
        .get('/?name=john')
        .expect(200);
      expect(response).toMatchSnapshot();
    });

    it('req.body', async () => {
      @Controller('/')
      class TestController {
        @Post('/')
        @Validate({
          body: {
            first: Joi.string().required(),
            last: Joi.string().default('Smith'),
            role: Joi.string().uppercase(),
          },
        })
        async getTest(req: e.Request, res: e.Response) {
          expect(req.body).toEqual({
            first: 'john',
            role: 'ADMIN',
            last: 'Smith',
          });
          return 'lol';
        }
      }

      appTest.addClass(TestController);

      await appTest.build();
      const response = await supertest(appTest.app)
        .post('/?end=lol')
        .send({
          first: 'john',
          role: 'admin',
        })
        .expect(200);
      expect(response).toMatchSnapshot();
    });
  });
});
