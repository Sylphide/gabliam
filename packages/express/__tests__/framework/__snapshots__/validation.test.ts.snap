// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`express integration update req values req.body 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-QDkmAz0AG1J53zfLvlKHt8fCZ/o\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": Object {
      "first": "john",
      "role": "admin",
    },
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "POST",
  },
  "status": 200,
  "text": "lol",
}
`;

exports[`express integration update req values req.headers 1`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip, deflate",
  "connection": "close",
  "secret-header": "@@@@@@",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`express integration update req values req.headers 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "1",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"1-E/vXnD05Dl1lhaIeEf9ewZcM/ww\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "accept": "application/json",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "k",
}
`;

exports[`express integration update req values req.params 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "1",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"1-E/vXnD05Dl1lhaIeEf9ewZcM/ww\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "k",
}
`;

exports[`express integration update req values req.query 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "1",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"1-E/vXnD05Dl1lhaIeEf9ewZcM/ww\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "k",
}
`;

exports[`express integration validations req.body 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "156",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"9c-lb73tokNqpEcTzZIECgZmqLKScw\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": Object {
      "first": "john",
      "last": 123,
    },
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "POST",
  },
  "status": 400,
  "text": "{\\"statusCode\\":400,\\"error\\":\\"Bad Request\\",\\"message\\":\\"child /\\"last/\\" fails because [/\\"last/\\" must be a string]\\",\\"validation\\":{\\"source\\":\\"body\\",\\"keys\\":[\\"last\\"]}}",
}
`;

exports[`express integration validations req.headers 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "228",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"e4-9YBgGPA9onlm1wj0GlR/2HIQaBw\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "accept": "application/json",
    },
    "method": "GET",
  },
  "status": 400,
  "text": "{\\"statusCode\\":400,\\"error\\":\\"Bad Request\\",\\"message\\":\\"child /\\"accept/\\" fails because [/\\"accept/\\" with value /\\"application&#x2f;json/\\" fails to match the required pattern: /xml/]\\",\\"validation\\":{\\"source\\":\\"headers\\",\\"keys\\":[\\"accept\\"]}}",
}
`;

exports[`express integration validations req.params 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "193",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"c1-Ucg8TlfIGVQQWOvNPRCiHk8y61o\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 400,
  "text": "{\\"statusCode\\":400,\\"error\\":\\"Bad Request\\",\\"message\\":\\"child /\\"id/\\" fails because [/\\"id/\\" must only contain alpha-numeric and underscore characters]\\",\\"validation\\":{\\"source\\":\\"params\\",\\"keys\\":[\\"id\\"]}}",
}
`;

exports[`express integration validations req.query 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "122",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"7a-iJ+h5zB08gNswU0QtTTcpNELWzs\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 400,
  "text": "{\\"statusCode\\":400,\\"error\\":\\"Bad Request\\",\\"message\\":\\"/\\"end/\\" is not allowed\\",\\"validation\\":{\\"source\\":\\"query\\",\\"keys\\":[\\"end\\"]}}",
}
`;
